runtime/null_coalesce.h

```diff
@@ -8,79 +8,131 @@
-  return value ? null_coalesce<ReturnType>(*value, fallback) : impl_::perform_fallback<ReturnType>(fallback);
-}
-
-template<class ReturnType, class KeyType, class FallbackType, class = vk::enable_if_constructible<mixed, KeyType>>
-ReturnType null_coalesce(const string &str, const KeyType &key, const FallbackType &fallback) noexcept {
-  string value = str.get_value(key);
-  return value.empty() ? impl_::perform_fallback<ReturnType>(fallback) : std::move(value);
-}
-
-template<class ReturnType, class KeyType, class FallbackType, class = vk::enable_if_constructible<mixed, KeyType>>
-ReturnType null_coalesce(const mixed &v, const KeyType &key, const FallbackType &fallback) noexcept {
-  return v.is_string()
-         ? null_coalesce<ReturnType>(v.as_string(), key, fallback)
-         : null_coalesce<ReturnType>(v.get_value(key), fallback);
-}
-
-template<typename ReturnType, typename FallbackType>
-ReturnType null_coalesce(const mixed &v, const string &string_key, int64_t precomuted_hash, const FallbackType &fallback) noexcept {
-  return v.is_string()
-         ? null_coalesce<ReturnType>(v.as_string(), string_key, fallback)
-         : null_coalesce<ReturnType>(v.get_value(string_key, precomuted_hash), fallback);
-}
+template<typename ResultType>
+struct NullCoalesce {
+public:
+  template<class ValueType>
+  explicit NullCoalesce(ValueType &&value) noexcept {
+    set_result(std::forward<ValueType>(value));
+  }
+
+  template<class ValueType, class KeyType, class = vk::enable_if_constructible<mixed, KeyType>>
+  NullCoalesce(const array<ValueType> &arr, const KeyType &key) noexcept {
+    if (auto *value = arr.find_value(key)) {
+      set_result(*value);
+    }
+  }
+
+  template<class ValueType>
+  NullCoalesce(const array<ValueType> &arr, const string &string_key, int64_t precomuted_hash) noexcept {
+    if (auto *value = arr.find_value(string_key, precomuted_hash)) {
+      set_result(*value);
+    }
+  }
+
+  template<class KeyType, class = vk::enable_if_constructible<mixed, KeyType>>
+  NullCoalesce(const string &str, const KeyType &key) noexcept {
+    set_result(str, key);
+  }
+
+  template<class KeyType, class = vk::enable_if_constructible<mixed, KeyType>>
+  NullCoalesce(const mixed &v, const KeyType &key) noexcept {
+    if (v.is_string()) {
+      set_result(v.as_string(), key);
+    } else {
+      set_result(v.get_value(key));
+    }
+  }
+
+  NullCoalesce(const mixed &v, const string &string_key, int64_t precomuted_hash) noexcept {
+    if (v.is_string()) {
+      set_result(v.as_string(), string_key);
+    } else {
+      set_result(v.get_value(string_key, precomuted_hash));
+    }
+  }
+
+  template<class FallbackType>
+  ResultType finalize(FallbackType &&fallback) noexcept {
+    return result_ ? std::move(*result_) : impl_::perform_fallback<ResultType>(std::forward<FallbackType>(fallback));
+  }
+
+  ~NullCoalesce() noexcept {
+    if (result_) {
+      result_->~ResultType();
```

Comment:
Shouldn't we assign `result_ = nullptr` here?

Reply:
No, this is a destructor, therefore result_ can't be used after it.